src/liboslexec/context.cpp

```diff
@@ -76,11 +76,13 @@ ShadingContext::execute_init (ShaderGroup &sgroup, ShaderGlobals &ssg, bool run)
     if (sgroup.nlayers()) {
         sgroup.start_running ();
         if (! sgroup.optimized()) {
-            shadingsys().optimize_group (sgroup);
+            auto ctx = shadingsys().get_context(thread_info());
+            shadingsys().optimize_group (sgroup, ctx);
```

Comment:
Could we use `*this` as the context?

Reply:
I don't think that's safe. Any setup that has happened before the execute_init could get clobbered. The "scratch context" that's needed for optimization can't be the same one that's already half set up to execute the shader.

Reply:
I did consider that. I just had trouble convincing myself that it was fully safe, and would not be super brittle (like, working fine now, but any number of future changes by unsuspecting developer/me would re-break it). Seemed safer to just request a fresh context. It's a pool, so it's very inexpensive to keep grabbing and then returning a simple scratch context.

Reply:
Makes sense.

LGTM!